src/if-run/builtins/sci-embodied/index.ts

```diff
@@ -1,190 +1,175 @@
+
+    return validate<z.infer<typeof schema>>(
+      schema as ZodType<any>,
+      mappedConfig
     );
   };
 
   /**
-   * Checks for required fields in input.
+   * Validates single observation for safe calculation.
    */
   const validateInput = (input: PluginParams) => {
-    const commonSchemaPart = (errorMessage: (unit: string) => string) => ({
-      'device/emissions-embodied': z
-        .number({
-          invalid_type_error: errorMessage('gCO2e'),
-        })
-        .gte(0)
-        .min(0),
-      'device/expected-lifespan': z
-        .number({
-          invalid_type_error: errorMessage('gCO2e'),
-        })
-        .gte(0)
-        .min(0),
-      duration: z
-        .number({
-          invalid_type_error: errorMessage('seconds'),
-        })
-        .gte(1),
+    const schema = z.object({
+      duration: z.number().gt(0),
+      vCPUs: z.number().gt(0).default(1),
+      memory: z.number().gt(0).default(16),
+      ssd: z.number().gte(0).default(0),
+      hdd: z.number().gte(0).default(0),
+      gpu: z.number().gte(0).default(0),
+      'usage-ratio': z.number().gt(0).default(1),
+      time: z.number().gt(0).optional(),
     });
 
-    const vcpusSchemaPart = {
-      'vcpus-allocated': z
-        .number({
-          invalid_type_error: SCI_EMBODIED_ERROR('count'),
-        })
-        .gte(0)
-        .min(0),
-      'vcpus-total': z
-        .number({
-          invalid_type_error: SCI_EMBODIED_ERROR('count'),
-        })
-        .gte(0)
-        .min(0),
-    };
-
-    const resourcesSchemaPart = {
-      'resources-reserved': z
-        .number({
-          invalid_type_error: SCI_EMBODIED_ERROR('count'),
-        })
-        .gte(0)
-        .min(0),
-      'resources-total': z
-        .number({
-          invalid_type_error: SCI_EMBODIED_ERROR('count'),
-        })
-        .gte(0)
-        .min(0),
-    };
+    return validate<z.infer<typeof schema>>(schema as ZodType<any>, input);
+  };
 
-    const schemaWithVcpus = z.object({
-      ...commonSchemaPart(SCI_EMBODIED_ERROR),
-      ...vcpusSchemaPart,
-    });
-    const schemaWithResources = z.object({
-      ...commonSchemaPart(SCI_EMBODIED_ERROR),
-      ...resourcesSchemaPart,
-    });
+  /**
+   * 1. Validates configuration and assigns defaults values if not provided.
+   * 2. Maps through observations and validates them.
+   * 3. Calculates total embodied carbon by substracting and the difference between baseline server and given one.
+   */
+  const execute = (inputs: PluginParams[]) => {
+    const safeConfig = validateConfig();
 
-    const schema = schemaWithVcpus.or(schemaWithResources).refine(allDefined, {
-      message: `All ${METRICS} should be present.`,
-    });
+    return inputs.map(input => {
+      const mappedInput = mapInputIfNeeded(input, mapping);
+      const safeInput = validateInput(mappedInput);
 
-    return validate<z.infer<typeof schema>>(schema, input);
+      const cpuE =
```

Comment:
move calculation to separate function

Reply:
there is no reason to move it, using safeConfig and safeInput object properties are making whole thing scary